src/ofonotracker.cpp

```diff
@@ -0,0 +1,335 @@
+private:
+    void evaluatePropertiesLater();
+    void getModemsAsync();
+    void getModemsSync();
+    void flushModems();
+    void setModems(const ModemInfoList &modemInfoList);
+    void addModem(const ModemInfo &modemInfo);
+    void removeModem(QString path);
+    void setModemProperty(QString path, QString key, QVariant val);
+    static bool trackedModemProperty(const QString &key);
+    OFonoTracker *q_ptr;
+    QDBusConnection m_systemBus;
+    QStringList m_modemPathList;
+    QHash<QString, ModemInfo> m_modemInfoHash;
+    QTimer *m_evaluatePropertiesTimer;
+    QStringList m_imeiNumbers;
+    Q_DISABLE_COPY(OFonoTrackerPrivate);
+    Q_DECLARE_PUBLIC(OFonoTracker);
+};
+
+OFonoTrackerPrivate::OFonoTrackerPrivate(bool synchronousInit, OFonoTracker *parent)
+    : QObject(parent)
+    , q_ptr(parent)
+    , m_systemBus(QDBusConnection::systemBus())
+    , m_evaluatePropertiesTimer(new QTimer(this))
+{
+    registerDBusTypes();
+    m_evaluatePropertiesTimer->setSingleShot(true);
+    m_evaluatePropertiesTimer->setInterval(20);
+    connect(m_evaluatePropertiesTimer, &QTimer::timeout, this, &OFonoTrackerPrivate::evaluateProperties);
+    m_systemBus.connect(OFONO_SERVICE, OFONO_MANAGER_OBJECT, OFONO_MANAGER_INTERFACE, OFONO_MANAGER_MODEM_ADDED,
+                        this, SLOT(handleManagerModemAdded(QDBusObjectPath, QVariantMap)));
+    m_systemBus.connect(OFONO_SERVICE, OFONO_MANAGER_OBJECT, OFONO_MANAGER_INTERFACE, OFONO_MANAGER_MODEM_REMOVED,
+                        this, SLOT(handleManagerModemRemoved(QDBusObjectPath)));
+    m_systemBus.connect(OFONO_SERVICE, QString(), OFONO_MODEM_INTERFACE, OFONO_MODEM_PROPERTY_CHANGED,
+                        this, SLOT(handleModemPropertyChanged(QString, QDBusVariant)));
+    /* For the benefit of existing code that assumes also D-Bus based values
+     * to be available immediately after the object is instantiated -> make
+     * it possible to utilize a synchronous probe and evaluate property values
+     * before returning from constructor. */
+    if (synchronousInit)
+        getModemsSync();
+    else
+        getModemsAsync();
+}
+
+OFonoTrackerPrivate::~OFonoTrackerPrivate()
+{
+}
+
+QStringList OFonoTrackerPrivate::imeiNumbers() const
+{
+    qWarning() << "#### call imeiNumbers() ->" << m_imeiNumbers;
+    return m_imeiNumbers;
+}
+
+void OFonoTrackerPrivate::evaluateProperties()
+{
+    qWarning() << "#### evaluate modem properties";
+    QStringList imeiNumbers;
+    for (auto iter = m_modemPathList.cbegin(); iter != m_modemPathList.cend(); ++iter) {
+        QString imei(m_modemInfoHash[*iter].m_properties.value(OFOFO_MODEM_PROPERTY_SERIAL).toString());
+        if (!imei.isEmpty())
+            imeiNumbers << imei;
+    }
+    if (m_imeiNumbers != imeiNumbers) {
+        m_imeiNumbers = imeiNumbers;
+        qWarning() << "#### emit imeiNumbersChanged() ->" << m_imeiNumbers;
+        Q_Q(OFonoTracker);
+        Q_EMIT q->imeiNumbersChanged();
```

Comment:
This `Q_EMIT` is also not used at all in this code base so I think `emit` would be more appropriate.

Reply:
will change